precompiles/common/precompiles.go

```diff
@@ -53,3 +55,22 @@ func AssertArgsLength(args []interface{}, length int) {
 		panic(fmt.Sprintf("expected %d arguments but got %d", length, len(args)))
 	}
 }
+
+func AssertNonPayable(value *big.Int) {
+	if value != nil && value.Sign() != 0 {
+		panic("sending funds to a non-payable function")
+	}
+}
+
+func HandlePayment(ctx sdk.Context, precompileAddr sdk.AccAddress, payer sdk.AccAddress, value *big.Int, bankKeeper BankKeeper) (sdk.Coin, error) {
+	usei, wei := state.SplitUseiWeiAmount(value)
+	if !wei.IsZero() {
+		return sdk.Coin{}, fmt.Errorf("selected precompile function does not allow payment with non-zero wei remainder: received %s", value)
+	}
+	coin := sdk.NewCoin(sdk.MustGetBaseDenom(), usei)
+	// refund payer because the following precompile logic will debit the payments from payer's account
+	if err := bankKeeper.SendCoins(ctx, precompileAddr, payer, sdk.NewCoins(coin)); err != nil {
```

Comment:
hm this seems to send coins from the precompile to the payer. Where is the logic that sends the coins from payer to precompile?